Plugin preload

corbosman opened the pull request:
I think this may be a solution for allowing plugins to load before sessions.

Ive introduced an rcube_plugins variable name $preload, which tells the plugin code to load this plugin early. Default is false, and then it gets loaded as it is now. If you set $preload in your plugin, you accept that all kinds of variables are not yet initialised, including $task and $app->output. 

I also had to account for $force, so ive also introduced a variable named $initialised, which tells the second phase of plugin init which plugins were already initialised in round 1. 

I think my logic with regards to noajax/noframe etc is ok, but please check. 

This PR would replace PR roundcube/roundcubemail#125

Since I need this code in my PR roundcube/roundcubemail#257 please let me know if you prefer this PR separate or if I should add it there since it's kind of related.

Commit:
Load plugins before sessions have started

Move the plugin loading phase to before sessions have started allowing plugins to add session drivers.  Plugins that want to use this should define an "onload" method in their plugins. This method does not have access to variables like $task as they are not yet initialised at that time.

corbosman commented:
I see what you mean now.. :)

I think we need to unfortunately add a parameter to load_config.. 

something like:

```
public function load_plugin($plugin_name,  $force = false, $require = true, $delay=false);
```

And load->plugins() then uses $delay=true;  

I dont see another way without breaking the load_plugin() api call..

thomascube commented:
Hmm what about:
```php
public function load_plugin(...)
{
    ...
    if (method_exists($plugin, 'onload')) {
        $plugin->onload();
    }
    if ($this->initialized) {
        $plugin->init();
    }
    $this->plugins[$plugin_name] = $plugin;
   ...
}

public function init_plugins(...)
{
    ...
    $this->initialized = true;
}
```

corbosman commented:
Ohh, I see now what you mean. Yes of course, you said move it to plugin_api. I was still thinking in my head about per-plugin status. 

You mean, you keep an initialized variable just to remember we did init_plugins and all further plugin loads should call init itself. Yup, no problem.

corbosman commented:
I have implemented the onload method check, and also refactored that noajax/noframe check into its own function to prevent code duplication and to make it more readable.

thomascube commented:
Thanks a lot! One last thing ;-)

`rcube_plugin_api::init()` became almost empty while `init_plugins()` has become pretty much of a replacement for it. So why not restore `init()` to what it was before and just add the plugin initialization loop to it. Then call `init()` after `load_plugins()`.

Just to keep the change as simple as possible and existing methods and signatures intact.

corbosman commented:
Sure no problem :)

corbosman commented:
Ok, that should do it. Seems to work :)

alecpl commented on program/include/rcmail.php:
```diff
@@ -93,6 +93,10 @@ protected function startup()
         }
 
+        // load all configured plugins
+        $this->plugins->load_plugins((array)$this->config->get('plugins', array()),
```
You moved this code before $task is being set. So, filtering by task in load_plugin() will not work as expected.

corbosman replied:
Im not sure what you mean. Isnt that the whole point? 

Load plugins is called early, but with the understanding that if you use an onload() method it wont have access to $task and other variables.  Then initialisation of the plugins, where task is checked, is defered until a later step. So $plugin->init(); in load_plugin is actually not called during the load_plugins() phase of the plugin system initialisation. 

Then later in rcmail.php $this->plugins->init($this, $this->task); is called where plugins have access to $task. 

I probably just dont understand your comment, so please elaborate which code path is going to fail and i'll be happy to fix it if possible.

alecpl replied:
OK, my mistake. The code is not simple to follow here. Maybe too much magic in it. Still init_plugins() method can be removed.

corbosman replied:
Good catch, didnt notice that unnecessary method. Committed. 

I totally agree with you the code is not easy to follow, but there are so many dependancies that I couldnt figure out a better way to do this without having to continuously modify more and more of the code.

alecpl commented on program/lib/Roundcube/rcube_plugin.php:
```diff
@@ -59,7 +59,7 @@
      */
     public $noframe = false;
-
+    
```
Please, clean this up.

alecpl commented on program/lib/Roundcube/rcube_plugin_api.php:
```diff
@@ -235,6 +255,42 @@ public function load_plugin($plugin_name, $force = false, $require = true)
+     */
+    private function filter($plugin)
+    {
+        return (    ($plugin->noajax  && !(is_object($this->output) && $this->output->type == 'html') )
```
I'd like to see also the task filter moved here.

corbosman commented:
I removed the spurious space (phpstorm added that somehow), moved $task check to filter. Please let me know what you meant with the first comment.

alecpl commented:
I think this is now ready to merge. Can we have this PR rebased as one commit (instead of eight)? Also, there are still some small CS issues like no space after 'foreach' or redundant spaces in 'if'.

corbosman commented:
I have rebased it into one commit, and hopefully found all the space issues.
